### PR TITLE
SPI TFT for F4 boards on STM32 Hal

### PR DESCRIPTION
**Pull request by rhapsodyv**
### Requirements

STM32 F4 and F1 boards with SPI TFT. And fix pinsDebug for F1.

### Description

This PR add support for SPI TFTs on STM32F4 boards for STM32 hal. 

### Tests Done

- [x] TFT_COLOR_UI on F1 with touch
- [x] TFT_COLOR_UI on F4 with touch 

### Benefits

SPI TFT for F4 boards

### Related Issues

#20280

### CODE REVIEW

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp [hunk 1]
```diff
@@ -34,35 +34,25 @@ DMA_HandleTypeDef TFT_SPI::DMAtx;
...
   SPI_TypeDef *spiInstance;
 
-  #if PIN_EXISTS(TFT_RESET)
-    OUT_WRITE(TFT_RESET_PIN, HIGH);
-    HAL_Delay(100);
-  #endif
-
-  #if PIN_EXISTS(TFT_BACKLIGHT)
-    OUT_WRITE(TFT_BACKLIGHT_PIN, HIGH);
-  #endif
```
Comment (sjasonsmith) [inline Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Why are these no longer needed?

File: Marlin/src/HAL/STM32/pinsDebug.h [hunk 2]
```diff
@@ -217,58 +204,62 @@ bool pwm_status(const pin_t Ard_num) {
...
+        case 11 : SERIAL_ECHOPGM("ETH"); break;
+        case 12 : SERIAL_ECHOPGM("FSMC, SDIO, OTG"); break;
+        case 13 : SERIAL_ECHOPGM("DCMI"); break;
+        case 14 : SERIAL_ECHOPGM("unused (shouldn't see this)"); break;
+        case 15 : SERIAL_ECHOPGM("EVENTOUT"); break;
+      }
     }
-  }
+  #else
+    // TODO: F1 doesn't support changing pins function, so we need to check the function of the PIN and if it's enabled
```
Comment (sjasonsmith) [inline Marlin/src/HAL/STM32/pinsDebug.h:263]: Does this work in pinsDebug in the F1 HAL, or was this functionality never present?

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp [hunk 3]
```diff
@@ -34,35 +34,25 @@ DMA_HandleTypeDef TFT_SPI::DMAtx;
...
   OUT_WRITE(TFT_A0_PIN, HIGH);
   OUT_WRITE(TFT_CS_PIN, HIGH);
 
   if ((spiInstance = (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_SCK_PIN),  PinMap_SPI_SCLK)) == NP) return;
   if (spiInstance != (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_MOSI_PIN), PinMap_SPI_MOSI)) return;
 
-  #if PIN_EXISTS(TFT_MISO) && (TFT_MISO_PIN != TFT_MOSI_PIN)
-    if (spiInstance != (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_MISO_PIN), PinMap_SPI_MISO)) return;
+  #if PIN_EXISTS(TFT_MISO)
+    if (TFT_MISO_PIN != TFT_MOSI_PIN)
```
Comment (sjasonsmith) [inline Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Why did this change from compile-time check to runtime check?

Comment (sjasonsmith) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Also applies to additional checks later in the file.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp [hunk 4]
```diff
@@ -74,31 +64,50 @@ void TFT_SPI::Init() {
...
       __HAL_RCC_SPI1_CLK_ENABLE();
-      __HAL_RCC_DMA1_CLK_ENABLE();
+      #ifdef STM32F1xx
+        __HAL_RCC_DMA1_CLK_ENABLE();
+        DMAtx.Instance = DMA1_Channel3;
+      #elif defined(STM32F4xx)
+        __HAL_RCC_DMA2_CLK_ENABLE();
+        DMAtx.Instance = DMA2_Stream3;
+        DMAtx.Init.Channel = DMA_CHANNEL_3;
+      #endif
```
Comment (sjasonsmith) [inline Marlin/src/HAL/STM32/tft/tft_spi.cpp:83]: This will be a problem for MCUs other than F1 and F4.

We must remember that currently this HAL is used for F0, F1, F4, F7, and soon H7 MCUs.

Anywhere we must behave differently on each MCU family, we should add something like this to ensure it is not used with unimplemented hardware.
```
#else
  #error "Not implemented for this MCU"
#endif
```

You may be able to avoid doing this in many places with some macros, with something like this, although you only have three places right now so it might not be worth the trouble.

```
#ifdef STM32F1xx
  #define SETUP_DMA(dest, dma, channel) do { \
    __HAL_RCC_DMA##dma##_CLK_ENABLE(); \
    (dest).Instance = DMA##dma##_Channel##channel; \
  } while(0)
#elif defined(STM32F4xx)
  #define SETUP_DMA(dest, dma, channel) do { \
    __HAL_RCC_DMA##dma##_CLK_ENABLE(); \
    (dest).Instance = DMA##dma##_Stream##channel; \
    (dest).Init.Channel = DMA_CHANNEL_##channel; \
  } while(0)
#else
  #error "Not implemented for this MCU"
#endif

File: platformio.ini [hunk 5]
```diff
@@ -1351,6 +1351,26 @@ extra_scripts        = ${common.extra_scripts}
   buildroot/share/PlatformIO/scripts/stm32_bootloader.py
   buildroot/share/PlatformIO/scripts/mks_encrypt.py
 
+#
+# MKS Robin Pro V2
+#
+[env:mks_robin_pro2]
```
Comment (sjasonsmith) [inline platformio.ini:1357]: Was the new board supposed to be in a different PR? It was not mentioned in the description.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 1 above)
Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Because it's handled by TFT_IO class now. The TFT code for STM32 was outdated compared with LPC and STM32F1.

File: Marlin/src/HAL/STM32/pinsDebug.h (hunk 2 above)
Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/pinsDebug.h:263]: It's not implemented for maple too, looks like it goes to a generic pinsDebug.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 3 above)
Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: because pins (Pax, PBx) in F4 are `static uint_8`, they aren't macro... so we can use compile time `#if` for compare pins on F4

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 4 above)
Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:83]: We can do it, but I don't think it worth do it now, for 2 reasons: all those SPI code will move to MarlinSPI soon, and enabling a TFT for others MCU may break code before that. Even the board needs specific pins set, and only F4 and F1 boards have it set now. So the build will stop before get this point.

File: platformio.ini (hunk 5 above)
Comment (rhapsodyv) [reply platformio.ini:1357]: I didn't put it in a separated PR because this a never-will-be-released board used only for development purposes. I pushed it because (1) I used it for F4 development and (2) because it is currently the first start point for *any* F4 board with TFT support. (we may use it as reference for the new Anet).

File: Marlin/src/HAL/STM32/pinsDebug.h (hunk 2 above)
Comment (sjasonsmith) [reply Marlin/src/HAL/STM32/pinsDebug.h:263]: There is a `pwm_details`, in HAL/STM32F1, although it doesn't seem to do a lot.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 4 above)
Comment (sjasonsmith) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:83]: I think you should put an `#error` at least once in each file which will cause problems. This is an aid for any future development, to make sure files that need to be updated cannot be missed.

Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:83]: Maybe it's matter for a sanity check: TFT will only work on F1 and F4. The #error for dma/spi can be added in the MarlinSPI, I guess.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 3 above)
Comment (thinkyhead) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: When is a hat not a hat?
```cpp
TFT_MISO_PIN != TFT_MOSI_PIN
```

Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Mosi master send data to slave, miso slave send data to master. When they are different, we have full duplex comm, if they are equal, we have only 1 line of data transfer, so we need to handle send and receive in the same pin.

Comment (sjasonsmith) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: > we need to handle send and receive in the same pin.

Do you have an example of this. That sounds like "not SPI" to me.

I know some devices will have MOSI but no MISO. I am not aware of any mechanism to perform half-duplex I/O with SPI over one pin.

Comment (rhapsodyv) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Well... that is the original code from jmz... I just ported it to F4, keeping same logic. He must have his reasons for that....

I know that stm32duino needs to know and have a lot of checks for spi 1 line or spi 2 lines. There's some specific flags that need to be cleared in one or in the other case.

But I really don't have any TFT that don't have both miso/mosi. Maybe some generic spi TFT have only one these PINs...

That 2 line vs 1 line, is only handled here. The other hals don't do it... 

For me, I would keep only 2 lines support... but as the code is there, i don't think it hurts.

File: platformio.ini [hunk 6]
```diff
@@ -1351,6 +1351,26 @@ extra_scripts        = ${common.extra_scripts}
...
+platform      = ${common_stm32.platform}
+extends       = common_stm32
+build_flags   = ${common_stm32.build_flags} -DHAL_HCD_MODULE_ENABLED  -DUSBHOST -DARDUINO_BLACK_F407VE
```
Comment (sjasonsmith) [inline platformio.ini:1360]: > `-DARDUINO_BLACK_F407VE`

Does this serve a purpose? This looks like something accidentally copied from `STM32F407VE_black`.

Comment (rhapsodyv) [reply platformio.ini:1360]: I’m using it just to be able to use the generic variant. It’s ok for testing purposes of most of the board features.

File: Marlin/src/HAL/STM32/tft/tft_spi.cpp (hunk 3 above)
Comment (sjasonsmith) [reply Marlin/src/HAL/STM32/tft/tft_spi.cpp:44]: Since it was already there, I think we can leave it here. It is possible that the STM32 SPI controllers support tri-stating the MOSI lines so the line could be shared, but more research would be needed to confirm or deny that.

I tend to think it is not useful unless we can find devices that we know use it. We could consider removing it in a future update.

**Comment by thisiskeithb** [comment]
Was the MKS Robin Pro V2 ever manufactured? I can only find info on the [Robin Pro V1 board](https://github.com/makerbase-mks/MKS-Robin-Pro) which is missing from Marlin.
